src/chainstate/stacks/index/trie_sql.rs

```diff
@@ -0,0 +1,293 @@
+
+ This file is part of Blockstack.
+
+ Blockstack is free software. You may redistribute or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License or
+ (at your option) any later version.
+
+ Blockstack is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY, including without the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::fmt;
+use std::error;
+use std::io;
+use std::io::{
+    Read,
+    Write,
+    Seek,
+    SeekFrom,
+    Cursor,
+    BufWriter,
+};
+
+use std::char::from_digit;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::collections::{VecDeque, HashMap, HashSet};
+
+use std::fs;
+use std::path::{
+    Path,
+    PathBuf
+};
+
+use std::os;
+use std::iter::FromIterator;
+
+use regex::Regex;
+
+use chainstate::burn::BlockHeaderHash;
+use chainstate::burn::BLOCK_HEADER_HASH_ENCODED_SIZE;
+
+use chainstate::stacks::index::{
+    TrieHash,
+    TRIEHASH_ENCODED_SIZE,
+    BlockMap,
+    trie_sql
+};
+
+use chainstate::stacks::index::storage::{
+    TrieFileStorage,
+};
+
+use chainstate::stacks::index::bits::{
+    get_node_byte_len,
+    write_nodetype_bytes,
+    read_hash_bytes,
+    read_block_identifier,
+    read_node_hash_bytes as bits_read_node_hash_bytes,
+    read_nodetype,
+    get_node_hash,
+};
+
+use chainstate::stacks::index::node::{
+    is_backptr,
+    clear_backptr,
+    set_backptr,
+    TrieNodeType,
+    TrieNode4,
+    TrieNode16,
+    TrieNode48,
+    TrieNode256,
+    TrieLeaf,
+    TrieNodeID,
+    TriePtr,
+    TriePath,
+    TrieNode
+};
+
+use rusqlite::{
+    Connection, OptionalExtension,
+    types::{ FromSql,
+             ToSql },
+    NO_PARAMS,
+    Error as SqliteError
+};
+
+use std::convert::{
+    TryFrom,
+    TryInto
+};
+
+use chainstate::stacks::index::Error as Error;
+
+use util::log;
+
+static SQL_MARF_DATA_TABLE: &str = "
+CREATE TABLE IF NOT EXISTS marf_data (
+   block_id INTEGER PRIMARY KEY, 
+   block_hash TEXT UNIQUE NOT NULL,
+   data BLOB NOT NULL
+);
+
+CREATE INDEX IF NOT EXISTS block_hash_marf_data ON marf_data(block_hash);
+";
+static SQL_MARF_MINED_TABLE: &str = "
+CREATE TABLE IF NOT EXISTS mined_blocks (
+   block_id INTEGER PRIMARY KEY, 
+   block_hash TEXT UNIQUE NOT NULL,
+   data BLOB NOT NULL
+);
+
+CREATE INDEX IF NOT EXISTS block_hash_mined_blocks ON mined_blocks(block_hash);
+";
+static SQL_CHAIN_TIPS_TABLE: &str = "
+CREATE TABLE IF NOT EXISTS chain_tips (block_hash TEXT UNIQUE NOT NULL);
+
+CREATE INDEX IF NOT EXISTS block_hash_chain_tips ON chain_tips(block_hash);
+";
+static SQL_EXTENSION_LOCKS_TABLE: &str = "
+CREATE TABLE IF NOT EXISTS block_extension_locks (block_hash TEXT UNIQUE NOT NULL);
```

Comment:
Is there a reason why `block_hash` isn't also the `PRIMARY KEY`?  Doing so ought to remove the need for the `block_hash_locks` index below.

Reply:
Yes -- I want block_id to also be the row_id, so that the blob can be opened directly from the block_id. SQLite only guarantees this if there is a single integer primary key.

Reply:
But `block_extension_locks` doesn't have any blobs in it?  I can see why this is needed for `marf_data` and `mined_blocks`, but not `block_extension_locks`?

Reply:
Oh, oops, I thought this comment was for the marf_data table -- yeah, I can make block_hash the primary key.

Reply:
Updated in c0fb0418d10800e764f88694b24e12ef917cdf39